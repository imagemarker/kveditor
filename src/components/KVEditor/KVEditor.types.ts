import { ReactElement } from 'react';

export interface KVEditConfigType {
  typeNotation?: string;
  nested?: boolean;
  validateKey?: RegExp;
  theme?: string;
}
export interface KVEditorPropsType {
  items?: KVItemType[];
  options?: KVEditConfigType;
  onChange?: (state: any) => void;
}
export interface KVEditorType {
  (props: KVEditorPropsType): ReactElement;
}
export interface KVItemType {
  key: string;
  value: any;
}
