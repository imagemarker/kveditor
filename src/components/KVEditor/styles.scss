.kv-editor {
  font-family: 'Source Sans Pro', sans-serif;
  font-weight: 400;
  margin: 3px;
  padding: 3px;
  border-radius: 3px;
  display: flex;
  flex-flow: column nowrap;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  min-width: 190px;

  &.light {
    color: #333;
    background-color: #FFF;
  }
  &.dark {
    color: #EEE;
    background-color: #444;
  }
}
