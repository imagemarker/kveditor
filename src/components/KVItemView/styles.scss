.kv-item-view {
  display: flex;
  flex-flow: row nowrap;
  margin-bottom: 3px;
  flex: 1;

  .key-value {
    flex: 1 1 auto;
    padding: 3px;
    border-radius: 3px;
    display: flex;
    flex-flow: row nowrap;
    min-width: 0;
    width: 100%;

    .key {
      font-weight: 600;
      padding: 3px 5px;
      margin-right: 3px;
      border-radius: 3px 0 0 3px;
    }
    .value {
      outline: none;
      flex: 1 1 auto;
      box-sizing: border-box;
      width: 100%;
    }
  }
  .remove, .save {
    flex: 0 0 auto;
    cursor: pointer;
    border-radius: 3px;
    padding: 4px;
    width: 32px;
    margin-left: 3px;
  }

  &.light {
    .key-value {
      background-color: #FFF;
      border: solid 1px #DDD;

      .key {
        background-color: #888;
        color: white;
      }
      .value {
        border: solid 1px #FFF;
        background-color: #FFF;
        color: #333;
      }
      &.key-id, &.key-disabled {
        background-color: #EEE;

        .value {
          border: solid 1px #EEE;
          background-color: #EEE;
        }
      }
      &.key-id {
        .key {
          background-color: #A00;
        }
      }
    }
    .remove {
      background: #EEE url('icons/bin-light.svg') no-repeat center center;
    }
    .save {
      background: #EEE url('icons/checkmark-light.svg') no-repeat center center;
    }
    .remove, .save {
      border: solid 1px #DDD;
      color: #333;

      &:hover {
        background-color: #CCC;
      }
    }
  }
  &.dark {
    .key-value {
      background-color: #888;
      border: solid 1px #333;

      .key {
        background-color: #EEE;
        color: #222;
      }
      .value {
        border: solid 1px #888;
        background-color: #888;
        color: #FFF;
      }
      &.key-id, &.key-disabled {
        background-color: #666;

        .value {
          border: solid 1px #666;
          background-color: #666;
        }
      }
      &.key-id {
        .key {
          background-color: #F99;
        }
      }
    }
    .remove {
      background: #666 url('icons/bin-dark.svg') no-repeat center center;
    }
    .save {
      background: #666 url('icons/checkmark-dark.svg') no-repeat center center;
    }
    .remove, .save {
      border: solid 1px #333;
      color: #CCC;

      &:hover {
        background-color: #555;
      }
    }
  }
}
