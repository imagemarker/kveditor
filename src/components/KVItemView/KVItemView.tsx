import React, { useState } from 'react';
import { KVItemViewType } from './KVItemView.types';
import './styles.scss';

const KVItemView: KVItemViewType = ({ item, dispatch, editorOptions }) => {
  const { key, value } = item;
  const [currentValue, setCurrentValue] = useState(value || '');
  const [isEditing, setEditing] = useState(false);

  const onRemove = () => {
    dispatch({ type: 'REMOVE', key });
  };

  const onChange = (event: React.FormEvent<HTMLInputElement>) => {
    setCurrentValue(event.currentTarget.value);
    setEditing(true);
  };

  const onSubmit = (event: React.FormEvent) => {
    event.preventDefault();
    dispatch({ type: 'UPDATE', item: { key, value: currentValue } });
    setEditing(false);
  };

  return (
    <form onSubmit={onSubmit} className={`kv-item-view ${editorOptions.theme}`}>
      <div className="key-value">
        <div className="key">{key}</div>
        <input className="value" name="value" type="text" value={currentValue} onChange={onChange} />
      </div>
      {isEditing ? <input className="save" type="submit" value="" title="Save" /> : null}
      <input className="remove" type="button" onClick={onRemove} title="Remove" />
    </form>
  );
};

export default KVItemView;
