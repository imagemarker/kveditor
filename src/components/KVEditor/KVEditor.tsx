import React, { useEffect, useReducer } from 'react';
import { KVEditConfigType, KVEditorType, KVItemType } from '@base/types';
import KVReducer from '../../hooks/KVReducer';
import KVItemEdit from '../KVItemEdit/KVItemEdit';
import KVItemView from '../KVItemView/KVItemView';
import './styles.scss';

const KVEditor: KVEditorType = ({ items = [], options, onChange }) => {
  const defaults = {
    theme: 'light',
    validateKey: new RegExp(/^[a-zA-Z][a-zA-Z0-9]*$/),
    typeNotation: 'string',
    nested: false
  };
  const editorOptions: KVEditConfigType = { ...defaults, ...options };
  const initialState = { items: items, keys: items.map(({ key }) => key) };
  const [state, dispatch] = useReducer(KVReducer, initialState);

  useEffect(() => {
    const items = state.items.reduce(
      (acc: Record<string, unknown>, { key, value }) => ({ ...acc, [key]: value }),
      {}
    );
    onChange && onChange(items);
  }, [onChange, state]);

  return (
    <div className={`kv-editor ${options.theme}`}>
      {state.items.map((item: KVItemType) => (
        <KVItemView
          key={item.key}
          item={item}
          dispatch={dispatch}
          editorOptions={editorOptions}
          viewOptions={null}
        />
      ))}
      <KVItemEdit keys={state.keys} dispatch={dispatch} editorOptions={editorOptions} />
    </div>
  );
};

export default KVEditor;
