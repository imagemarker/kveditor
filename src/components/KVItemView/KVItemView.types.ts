import { Dispatch, Reducer, ReducerAction } from 'react';
import { KVEditConfigType, KVItemType } from '@base/src/components/KVEditor/KVEditor.types';
import { KVAction, KVState } from '@base/src/hooks/KVReducer.types';

export interface KVItemViewPropsType {
  dispatch: Dispatch<ReducerAction<Reducer<KVState, KVAction>>>;
  item: KVItemType;
  editorOptions: KVEditConfigType;
  viewOptions: any;
}
export interface KVItemViewType {
  (props: KVItemViewPropsType): JSX.Element;
}
