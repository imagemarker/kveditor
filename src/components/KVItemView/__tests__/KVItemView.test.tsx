import React from 'react';
import { render } from '@testing-library/react';
import { KVEditConfigType } from '../../KVEditor/KVEditor.types';
import KVItemView from '../KVItemView';

describe('KVItemView', () => {
  const validateKey = new RegExp(/^[a-zA-Z][a-zA-Z0-9]*$/);
  const dispatch = jest.fn();

  const tree = (options: KVEditConfigType = { theme: 'light', validateKey }) =>
    render(
      <KVItemView key="key" item={{ key: 'foo', value: 'bar' }} dispatch={dispatch} editorOptions={options} />
    );

  test('renders properly', () => {
    const { asFragment } = tree();
    expect(asFragment()).toMatchSnapshot();
  });
});
