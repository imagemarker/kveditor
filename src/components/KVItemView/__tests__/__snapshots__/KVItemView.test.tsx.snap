// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KVItemView renders properly 1`] = `
<DocumentFragment>
  <form
    class="kv-item-view light"
  >
    <div
      class="key-value"
    >
      <div
        class="key"
      >
        foo
      </div>
      <input
        class="value"
        name="value"
        type="text"
        value="bar"
      />
    </div>
    <input
      class="remove"
      title="Remove"
      type="button"
    />
  </form>
</DocumentFragment>
`;
